include/oneapi/dpl/pstl/experimental/internal/for_loop_impl.h

```diff
@@ -476,53 +476,22 @@ __pattern_for_loop(_ExecutionPolicy&& __exec, _Ip __first, _Ip __last, _Function
 template <typename _Ip>
 struct __use_par_vec_helper
 {
-  protected:
-
-    template <typename _Tag>
-    static constexpr auto __get_is_vector(_Tag)
-    {
-        return typename _Tag::__is_vector{};
-    }
-
-#if _ONEDPL_BACKEND_SYCL
-    template <typename _BackendTag>
-    static constexpr auto __get_is_vector(__hetero_tag<_BackendTag>)
-    {
-        return ::std::true_type{};
-    }
-#endif // _ONEDPL_BACKEND_SYCL
-
-    template <typename _Tag>
-    static constexpr auto __get_is_parallel(_Tag)
-    {
-        return oneapi::dpl::__internal::__is_parallel_tag<_Tag>{};
-    }
-
-#if _ONEDPL_BACKEND_SYCL
-    template <typename _BackendTag>
-    static constexpr auto __get_is_parallel(__hetero_tag<_BackendTag>)
-    {
-        return ::std::true_type{};
-    }
-#endif // _ONEDPL_BACKEND_SYCL
-
-  public:
     using __it_type = std::conditional_t<std::is_integral_v<_Ip>, _Ip*, _Ip>;
 
     template <typename _ExecutionPolicy>
     static constexpr auto
     __use_vector(_ExecutionPolicy&& __exec)
     {
         using __tag_type = decltype(oneapi::dpl::__internal::__select_backend(__exec, std::declval<__it_type>()));
-        return __get_is_vector(__tag_type{});
+        return typename __tag_type::__is_vector{};
     }
 
     template <typename _ExecutionPolicy>
     static constexpr auto
     __use_parallel(_ExecutionPolicy&& __exec)
     {
         using __tag_type = decltype(oneapi::dpl::__internal::__select_backend(__exec, std::declval<__it_type>()));
-        return __get_is_parallel(__tag_type{});
+        return oneapi::dpl::__internal::__is_parallel_tag<__tag_type>{};
```

Comment:
If you mean the state of code from https://github.com/oneapi-src/oneDPL/tree/dev/skopienko/tag_dispatching then I think it's not important.
But also we have `return typename _Tag::__is_vector{};` in tag dispatching code which also absent in hetero tags.

Reply:
@SergeyKopienko, to tell the truth I did not get your answear.
My question is "why do wee need oneapi::dpl::__internal::__is_parallel_tag?"
Because we can use `is_same` stuff directly in the mentioned line where we need to check a tag type. Even we want to keep `__is_parallel_tag ` stuff, why  did we place it in `execution_impl.h`?  `__is_parallel_tag ` is not used everywhere expeting for_loop_impl.h and just one line.

Reply:
Initially it was the comment from @rarutyun :
"
In [include/oneapi/dpl/pstl/experimental/internal/for_loop_impl.h](https://github.com/oneapi-src/oneDPL/pull/1239#discussion_r1522102949):
>      {
-        return oneapi::dpl::__internal::__is_parallelization_preferred<_ExecutionPolicy, _Ip>(
-            ::std::forward<_ExecutionPolicy>(__exec));
+        using __tag_type = decltype(oneapi::dpl::__internal::__select_backend(__exec, std::declval<__it_type>()));
+        using __is_parallel_tag = std::disjunction<
seems like it can be done even simple because we already have __is_parallel_tag, implemented in this PR. Let's use that.
"

Reply:
And `__is_parallel_tag` directly or indirectly used more then one place of our code.

Reply:
Hope now you have my full answer about `__is_parallel_tag` and all it's usages.